debitpolicy/debitpolicy.go

```diff
@@ -0,0 +1,115 @@
+	switch x := m.(type) {
+	case events.AccountDebitedForWithdrawal:
+		p.DebitAmount += x.Amount
+	case events.AccountDebitedForTransfer:
+		p.DebitAmount += x.Amount
+	}
+}
+
+// Aggregate implements the business logic for a debit policy.
+//
+// It centralizes all debits that are applied to an account over a period of
+// time in order to enforce a limited debit policy.
+type Aggregate struct{}
+
+// New returns a new debit policy instance.
+func (Aggregate) New() dogma.AggregateRoot {
+	return &debitpolicy{}
+}
+
+// Configure configures the behaviour of the engine as it relates to this
+// handlers.
+func (Aggregate) Configure(c dogma.AggregateConfigurer) {
+	c.Name("debitpolicy")
+	c.RouteCommandType(commands.CheckWithdrawalAllowedByDebitPolicy{})
+	c.RouteCommandType(commands.CheckTransferAllowedByDebitPolicy{})
+}
+
+// RouteCommandToInstance returns the ID of the aggregate instance that is
+// targetted by m.
+func (Aggregate) RouteCommandToInstance(m dogma.Message) string {
+	switch x := m.(type) {
+	case commands.CheckWithdrawalAllowedByDebitPolicy:
+		return makeInstanceID(x.Timestamp, x.AccountID)
+	case commands.CheckTransferAllowedByDebitPolicy:
+		return makeInstanceID(x.Timestamp, x.AccountID)
+	default:
+		panic(dogma.UnexpectedMessage)
+	}
+}
+
+// HandleCommand handles a command message that has bene routed to this handler.
+func (Aggregate) HandleCommand(s dogma.AggregateCommandScope, m dogma.Message) {
+	if s.Create() {
+		s.Log("debit policy created")
+	}
+
+	switch x := m.(type) {
+	case commands.CheckWithdrawalAllowedByDebitPolicy:
+		checkWithdrawalAllowed(s, x)
+	case commands.CheckTransferAllowedByDebitPolicy:
+		checkTransferAllowed(s, x)
+	default:
+		panic(dogma.UnexpectedMessage)
+	}
+}
+
+func checkWithdrawalAllowed(s dogma.AggregateCommandScope, m commands.CheckWithdrawalAllowedByDebitPolicy) {
+	p := s.Root().(*debitpolicy)
+
+	if p.DebitAmount+m.Amount <= debitLimitPerPeriod {
+		s.RecordEvent(events.WithdrawalApprovedByDebitPolicy{
+			TransactionID: m.TransactionID,
+			AccountID:     m.AccountID,
+			Amount:        m.Amount,
+		})
+	} else {
+		s.RecordEvent(events.WithdrawalDeclinedByDebitPolicy{
+			TransactionID: m.TransactionID,
+			AccountID:     m.AccountID,
+			Amount:        m.Amount,
+		})
+	}
+}
+
+func checkTransferAllowed(s dogma.AggregateCommandScope, m commands.CheckTransferAllowedByDebitPolicy) {
+	p := s.Root().(*debitpolicy)
+
+	if p.DebitAmount+m.Amount <= debitLimitPerPeriod {
+		s.RecordEvent(events.TransferApprovedByDebitPolicy{
+			TransactionID: m.TransactionID,
+			AccountID:     m.AccountID,
+			Amount:        m.Amount,
+		})
+	} else {
+		s.RecordEvent(events.TransferDeclinedByDebitPolicy{
+			TransactionID: m.TransactionID,
+			AccountID:     m.AccountID,
+			Amount:        m.Amount,
+		})
+	}
+}
+
+func makeInstanceID(t time.Time, accountID string) string {
+	return fmt.Sprintf("%d%d%d-%s", t.Year(), t.Month(), t.Day(), accountID)
```

Comment:
Perhaps some padding on day/month here.

Reply:
Good catch. I struggled a bit trying to find date formatting options for Go and forgot to come back and fix it.

Reply:
Take a look at this: https://golang.org/pkg/time/#pkg-constants

Time formatting in Go is a little unusual. There is a "reference time" of "Mon Jan 2 15:04:05 MST 2006", which provides the values you use to format a time.

If for example, you wanted to format an ISO date (`YYYY-MM-DD`) you would use:

```go
t := time.Now()
t.Format("2006-01-02") // Jan 2, 2006, as per the reference time.
```

There are a bunch of constants for time formats (for example `time.RFC3339`), but none predefined for ISO dates as far as I can tell.

Reply:
Ok, i saw the Go date formatting thing and got really confused. So that "reference time" is some kind of magic field marker thing that is used to identify where to put each value of a different given time to be formatted?

Reply:
Yeah, I think it's just that every field has a different numeric value, so they know if they see a `1` or an `01`, etc it's always talking about the month. I guess it's designed such that the format string looks like the date it will will produce, but IMO it's MUCH harder to remember the values to use than it is to imagine what `YYYY-MM-DD` would look like :P but it works well enough.

Reply:
The issue Ii see with the magic number formatting is you can't easily jam the numbers together? I suppose none would overlap, but it just feels unsafe.
I wanted to do a compact "YYYYMMDD-AccountID", but maybe i should still keep dashes between period types?